iree/compiler/Dialect/Flow/Transforms/HLOToHLOPreprocessing.cpp

```diff
@@ -171,54 +229,132 @@ class ExtractConvOpPaddingAttributes : public OpRewritePattern<mhlo::ConvOp> {
   }
 };
 
-struct OrderConvFeatureDimensions : public OpRewritePattern<mhlo::ConvOp> {
+// Guarantee that the input dimensions are ordered batch, spatial_dims, feature
+// dim.
+class ReorderConvOpInputDimensions : public OpRewritePattern<mhlo::ConvOp> {
```

Comment:
Would this make more sense as a canonicalization on mhlo::ConvOp upstream? Or a structural change to mhlo to just order them consistently wherever they are created?

Reply:
> Would this make more sense as a canonicalization on mhlo::ConvOp upstream?

Personally I would think it does however technically that goes against how the main
XLA backend works. Because they generate using the convolution attributes these kind of
decompositions are usually discouraged.

>  Or a structural change to mhlo to just order them consistently wherever they are created?

If this were only a matter of a couple locations, this would be fine, however we can get
XLA from a couple different sources (mlir xla-builder, xla proto, TF legalization) and
guaranteeing only a single form would be difficult to guarantee.

If the XLA legalization path agrees on a similar canonical representation I would be
inclined to upstream, however thats extremely unlikely.